src/Features/Core/Portable/FindUsages/DefinitionItem.DocumentLocationDefinitionItem.cs

```diff
@@ -11,25 +17,112 @@ internal partial class DefinitionItem
         /// <see cref="DocumentSpan"/>.
         /// </summary>
         // internal for testing purposes.
-        internal sealed class DocumentLocationDefinitionItem : DefinitionItem
+        internal sealed class DefaultDefinitionItem : DefinitionItem
         {
+            private readonly Workspace _workspaceOpt;
+
             internal override bool IsExternal => false;
 
-            public DocumentLocationDefinitionItem(
+            public DefaultDefinitionItem(
+                Workspace workspaceOpt,
                 ImmutableArray<string> tags,
                 ImmutableArray<TaggedText> displayParts,
                 ImmutableArray<TaggedText> nameDisplayParts,
+                ImmutableArray<TaggedText> originationParts,
                 ImmutableArray<DocumentSpan> sourceSpans,
                 ImmutableDictionary<string, string> properties,
                 bool displayIfNoReferences)
-                : base(tags, displayParts, nameDisplayParts,
-                       ImmutableArray.Create(new TaggedText(TextTags.Text, sourceSpans[0].Document.Project.Name)),
+                : base(tags, displayParts, nameDisplayParts, originationParts,
                        sourceSpans, properties, displayIfNoReferences)
             {
+                _workspaceOpt = workspaceOpt;
             }
 
-            public override bool CanNavigateTo() => SourceSpans[0].CanNavigateTo();
-            public override bool TryNavigateTo() => SourceSpans[0].TryNavigateTo();
+            public override bool CanNavigateTo()
+            {
+                if (this.Properties.ContainsKey(NonNavigable))
+                {
+                    return false;
+                }
+
+                if (this.Properties.TryGetValue(MetadataSymbolKey, out var symbolKey))
+                {
+                    return CanNavigateToMetadataSymbol(symbolKey);
+                }
+
+                return SourceSpans[0].CanNavigateTo();
+            }
+
+            public override bool TryNavigateTo()
+            {
+                if (this.Properties.ContainsKey(NonNavigable))
+                {
+                    return false;
+                }
+
+                if (this.Properties.TryGetValue(MetadataSymbolKey, out var symbolKey))
+                {
+                    return TryNavigateToMetadataSymbol(symbolKey);
+                }
+
+                return SourceSpans[0].TryNavigateTo();
+            }
+
+            private bool CanNavigateToMetadataSymbol(string symbolKey)
+                => TryNavigateToMetadataSymbol(symbolKey, (symbol, project, service) => true);
+
+            private bool TryNavigateToMetadataSymbol(string symbolKey)
+                => TryNavigateToMetadataSymbol(symbolKey, (symbol, project, service) =>
```

Comment:
💭 For me this exceeded the complexity of an expression for which expression-bodied members provides a readability benefit.